### DISCUSSION TITLE
Messages on queue with message TTL occasionally do not expire

### DISCUSSION THREAD

**Discussion by ppminfin** [rabbitmq/rabbitmq-server#10719] [Other] [answered]
### Describe the bug

Messages on queue with message TTL occasionally do not expire. When this happens a restart of the RabbitMQ server does not help. There are no consumers and the queue is defined with the following features and policy:

x-dead-letter-exchange
x-dead-letter-routing-key
x-message-ttl: 3600000
arguments: x-queue-type: quorum
durable: true

dead-letter-strategy: at-least-once
overflow: reject-publish
queue-leader-locator: balanced

### Reproduction steps

1. Publish a message to the queue
2. Wait for the TTL of 1 hour and the message remains stuck on the queue

The following image is showing the headers of the message at the head of the queue

![image](https://github.com/rabbitmq/rabbitmq-server/assets/163117445/87bdad02-c53b-44c5-b9f8-b673b0fe0c1c)

### Expected behavior

1. Publish a message to the queue
2. Wait for the TTL of 1 hour and the message expires and is routed to the x-dead-letter-exchange and x-dead-letter-routing-key

**Comment by michaelklishin** [comment, marked as answer]
This is documented and has been discussed here on GitHub many times before.

Messages expire [when they reach the head of the queue](https://www.rabbitmq.com/docs/ttl) and not earlier. Also see the section on retroactive application of a policy.

Consumers will never receive expired messages as deliveries, though, so when exactly messages are marked for deletion does not matter for any queue that has consumers at least most of the time.

Queues that do not have consumers for long time periods should be streams, where time-based expiration is the name of the game.
